### PR TITLE
RDM Melee Combo Refactoring & specific recommendation table

### PR DESCRIPTION
**Pull request by AzariahAeroborn**
The major goal of this set of changes is to add an output table for Melee combos, similar to the RotationTable that PLD FightOrFlight and Requiscat are using.  

To better support the new module, I refactored the Gauge module in three ways:
1) Changed the handling for mana changes to fabricate a new event type (rdmCast) that includes the before and after mana as well as mana loss due to overcap, imbalance, or miss/invulnerable.  
2) Exported some of the mana related constant arrays for other modules to use
3) Moved the melee combo analysis & suggestions into the new Melee Combos module.

Along with the move of the analysis & suggestions, I did also review Jump's MoogleMedia RDM guide (https://www.mooglemedia.com/rdm-guide/) and refactored the melee combo suggestions according to his suggestions for when to delay the combo and/or dump procs.  **There is a logic change here!**
- Reading through Jump's guide, and the notes in \parser\jobs\rdm\NOTES.md, the recommendation for when to delay the combo is not based on a hard guideline of 96 mana in the proc being dumped, but rather a guideline of do not lose more than 7 mana to overcapping if you're dumping procs.  Specifically, in the (Not So) Scary Math section, under Potency Per Mana, Jump summarizes "This means that converting a 20% proc chance to a 100% proc chance is effectively a ~34.8 potency gain.  You would need at least 8 Mana to overtake this difference (8 * 4.5 = 36)."  In light of that, I removed the MANA_DONT_CAST_THRESHOLD constant and replaced it with a IGNORE_FINISHER_PROCS_MANA_LOSS_THRESHOLD constant in the MeleeCombos module instead.
- This also made it so that the "bothprocsup" context for missed finishers that should have been a dumped proc instead changed, since the recommendation can also hold true in a situation like the first melee combo in pull 10 of this log: https://www.fflogs.com/reports/gwYpxjNnLPhF3C1r -- the melee combo started at 84|100 with VerStone Ready up and Acceleration nearly available, and over-wrote the VerStone Ready proc.  Delaying with VerStone + VerAero to hit 100|100 and then running the combo with a VerFlare finisher (and Acceleration) is guaranteed to generate a VerFire Ready proc (whether or not VerStone Ready procs from the VerAero used to delay) at only a cost of 4 mana overflow, making it a net gain per Jump's math.

The only other logic change I have is the tracking and new suggestion for "don't be a dumbass and attack the wrong Omega M/F".  I'll work on synthing a log with some misses in Eureka to ensure they're going to hit the same condition, and would be happy to accept suggestions for changes to the copy on that suggestion... it seems too wordy to me, but I haven't come up with anything more concise that I like.

I'll also create a dummy parse with the various test cases for melee finisher recommendations and reply with a comment here when that's created, it will probably be tomorrow though.

Oh, and I also fixed the spelling of VERAERO, because that was bugging the hell out of me while I was working on this.  Sorry, I should have probably split that out into its own commit so it could be merged separately from all this mess 😛

### CODE REVIEW

**Comment by AzariahAeroborn** [comment]
My other follow-up question is, what do I need to do to add the English strings to the correct localization place for the new translation elements (and to update the one I changed)?  I'd be happy to do it, just haven't figured out where to go to do that.

**Comment by AzariahAeroborn** [comment]
I have a set of test cases put together, though I didn't quite create all of the cases I meant to.  Log is here: https://www.fflogs.com/reports/nV94rkt6dNDBT83f

The first two pulls are in Eureka, to confirm that 
- Scatter is only gaining mana one time per cast
- Misses are being detected the same way as invuln due to firewall

In testing, I observed that Scatter was in fact calculating mana gain for each separate damage hit.  Changed event hook to aoedamage resolve.  Misses worked as expected.

I then synthesized melee combos on a striking dummy, based on the following table:
```
# |   Case XIVA URL Path  | Balance? | L Proc | H Proc | Accel | Del Loss | Expected
1 | nV94rkt6dNDBT83f/4/7/ |    No    |   N    |        |       |          | Use Low Finisher
2 | nV94rkt6dNDBT83f/3/7/ | Low<= -9 |   Y    |   N    |   Y   |          | Use High Finisher    
3 | nV94rkt6dNDBT83f/11/7/| Low<= -9 |   Y    |   N    |   N   | Low      | Delay
4 | nV94rkt6dNDBT83f/13/7/| Low<= -9 |   Y    |   N    |   N   | High     | Use High Finisher
5 | nV94rkt6dNDBT83f/6/7/ | Low > -9 |   Y    |   N    |       | Low      | Delay
6 | nV94rkt6dNDBT83f/14/7/| Low > -9 |   Y    |   N    |       | High     | Use Low Finisher
7 |                       |    No    |   Y    |   Y    |       | Low      | Delay w/Low Proc + Opposite
8 | nV94rkt6dNDBT83f/15/7/|    No    |   Y    |   Y    |       | High     | Use Low Finisher
9 | nV94rkt6dNDBT83f/10/7/|    Yes   |        |        |   Y   |          | Use Any Non-Proc Finisher
10|                       |    Yes   |   N    |   N    |   N   | Low      | Delay w/Impact or Jolt
11| nV94rkt6dNDBT83f/7/7/ |    Yes   |   N    |   N    |   N   | High     | Use Any Finisher
12| nV94rkt6dNDBT83f/18/7/|    Yes   |   Y    |   N    |   N   | Low      | Delay w/Proc + Opposite
13| nV94rkt6dNDBT83f/19/7/|    Yes   |   Y    |   N    |   N   | High     | Use Non-Proc Finisher
14|                       |    Yes   |   Y    |   Y    |       | Low      | Delay w/Proc + Opposite
15| nV94rkt6dNDBT83f/22/7/|    Yes   |   Y    |   Y    |       | High     | Use Any Finisher
```

I found and fixed issues with a couple of the combos, I still need to fix case 9, as that seems to not be detecting acceleration, as well as to generate correct test cases for 7, 10, and 14.  I'm going to commit the changes to cover the Scatter issue and the fixed cases tomorrow, I ran out of time to finish before raid tonight.

**Comment by AzariahAeroborn** [comment]
Added the fixes I made for combos that weren't displaying correctly, and for multi-hit scatter.  I believe this should be ready for review & merging.
cc: @ackwell @Cephid

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+				}
+				if (finisherUsed.guid === ACTIONS.VERFLARE.id) {
+					this._incorrectFinishers.verflare++
+				}
+				combo.finisher.recommendation = <Trans id="rdm.meleecombos.recommendation.incorrect">See the suggestions section for finisher guidelines.</Trans>
+			}
+		}
+	}
+
+	_outOfBalanceFinisher(lowerMana, lowerManaProc, lowerManaActions, higherMana, higherManaProc, higherManaActions) {
```
Comment (ackwell) [inline src/parser/jobs/rdm/modules/MeleeCombos.js]: few more args than i'm comfy with - could probably clean it up as like
```js
// Only need to do this once for each mana type
const whiteState = {
	amount: ...,
	proc: ...,
	actions: ...,
}

// Then can pass directly
this._outOfBalanceFinisher(whiteState, blackState) // or (blackState, whiteState)
```
can destructure the params on the receiving end

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+		if (combo.startmana.white < combo.startmana.black) {
+			recommendedFinisher = this._outOfBalanceFinisher(combo.startmana.white, isStoneReady, WHITE_MANA_ACTIONS, combo.startmana.black, isFireReady, BLACK_MANA_ACTIONS)
+		} else if (combo.startmana.black < combo.startmana.white) {
+			recommendedFinisher = this._outOfBalanceFinisher(combo.startmana.black, isFireReady, BLACK_MANA_ACTIONS, combo.startmana.white, isStoneReady, WHITE_MANA_ACTIONS)
+		} else {
+			recommendedFinisher = this._inBalanceFinisher(combo.startmana.black, isFireReady, BLACK_MANA_ACTIONS, combo.startmana.white, isStoneReady, WHITE_MANA_ACTIONS)
+		}
+
+		if (recommendedFinisher instanceof Array) {
+			if (recommendedFinisher === FINISHERS) {
```
Comment (ackwell) [inline src/parser/jobs/rdm/modules/MeleeCombos.js]: seems like a peculiar way of representing it, but i'll water under the bridge it 'cus i don't gotta maintain it :blobmorning:

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+
+				if (newHigherMana < newLowerMana) {
+					// Mana rebalancing resulted in the original higherMana becoming the lower total (guaranteed proc), valid option, push onto stack
+					possibleDelays.push({
+						finisher: [higherManaActions.proc, lowerManaActions.dualcast, higherManaActions.finisher],
+						manaLoss: manaLoss,
+					})
+				}
+				// Disregard case when newHigherMana remains even or higher than newLowerMana - error in calling the delay module
+				console.log(`Delay module called with ${lowerMana}|${higherMana} starting mana, no lowerManaProc, and a higherManaProc.`)
```
Comment (ackwell) [inline src/parser/jobs/rdm/modules/MeleeCombos.js]: missed some logging~
if it's a theoretically impossible case, it should be triggering `brokenLog` - if it's not, it can either be squashed or presented to the user in some manner.

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+										onClick={() => this.timeline.show(start, end)}
+									/>}
+								</Table.Cell>
+								<Table.Cell>
+									<span style={{whiteSpace: 'nowrap'}}>{white} White | {black} Black</span>
+								</Table.Cell>
+								<Table.Cell textAlign="center">
+									<span>{
+										Object.keys(combo.startprocs).map((key) => {
+											if (combo.startprocs[key]) {
```
Comment (ackwell) [inline src/parser/jobs/rdm/modules/MeleeCombos.js]:
```js
// prefer
if (!condition) { return }
...

// over
if (condition) {
	...
}
```

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 5]
```diff
@@ -44,30 +42,124 @@ export const SEVERITY_LOST_MANA = {
...
-const MANA_CAP = 100
+export const MANA_CAP = 100
 const ENHANCED_SCATTER_GAIN = 8
 const ENHANCED_SCATTER_44_GAIN = 10
-const MISSING_HARDCAST_MANA_VALUE = 11
-const MANAIFCATION_MULTIPLIER = 2
-const MANA_DONT_CAST_THRESHOLD = 96
-const FINISHER_GAIN = 21
-const MELEE_COMBO_COST = 80
+//const MISSING_HARDCAST_MANA_VALUE = 11  //removed with refactor to GaugeAction, change to damage type events prevents need for guessing at missing casts
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: Whoops, thought I got this.  Fixed now.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 6]
```diff
@@ -44,30 +42,124 @@ export const SEVERITY_LOST_MANA = {
...
+const MANAFICATION_MULTIPLIER = 2
+
+class GaugeAction {
+	mana = {
+		white: {
+			beforecast: 0,
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: Fixed.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 7]
```diff
@@ -44,30 +42,124 @@ export const SEVERITY_LOST_MANA = {
...
+	constructor(startingWhite, startingBlack) {
+		this.mana.white.beforecast = startingWhite
+		this.mana.black.beforecast = startingBlack
+
+		this.mana.white.aftercast = startingWhite
+		this.mana.black.aftercast = startingBlack
+	}
+
+	calculateManaFicationManaGained() {
+		//console.log(`White: ${this._whiteMana}, Black: ${this._blackMana}`)
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: This was copied from the pre-existing code when I moved things around, though I guess it's actually just flat out not correct now.  Removed.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 8]
```diff
@@ -80,247 +172,69 @@ export default class Gauge extends Module {
...
-
 		constructor(...args) {
 			super(...args)
 
-			this.addHook('cast', {by: 'player'}, this._onCast)
+			this.addHook('cast', {
+				by: 'player',
+				abilityId: ACTIONS.MANAFICATION.id,
+			}, this._onCast)
+			this.addHook('aoedamage', {by: 'player'}, this._onCast)
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: The alternate name I can come up with is _gaugeEvent 🤷‍♂️   I'm kind of ambivalent to renaming this or not.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 9]
```diff
@@ -80,247 +172,69 @@ export default class Gauge extends Module {
...
+				this._pushToGraph()
 			}
-			if (useVerHoly && abilityId === ACTIONS.VERFLARE.id) {
-				this._incorrectFinishers.verflare++
-			} else if (abilityId === ACTIONS.VERFLARE.id && useOnBadProc) {
-				this._incorrectFinishers.verflare++
+
+			const fabricatedEvent = {
+				...event,
+				type: 'rdmCast',
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: Changed.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 10]
```diff
@@ -370,6 +268,18 @@ export default class Gauge extends Module {
 				</Fragment>,
 			}))
 
+			this.suggestions.add(new TieredSuggestion({
+				icon: ACTIONS.VERHOLY.icon,
+				content: <Fragment>
+					<Trans id="rdm.gauge.suggestions.white-mana-invuln-content">Ensure you don't target a boss that you cannot damage (e.g. due to the Packet Filter debuffs in Omega M/F) with your damaging spells.  Spells that do no damage due to an invulnerable target or due to missing result in no mana gained, which potentially costs you one or more Enchanted Combos.</Trans>
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: Removed both callouts to the specific boss mechanic

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+	proc: ACTIONS.VERSTONE,
+	dualcast: ACTIONS.VERAERO,
+	finisher: ACTIONS.VERHOLY,
+}
+const BLACK_MANA_ACTIONS = {
+	proc: ACTIONS.VERFIRE,
+	dualcast: ACTIONS.VERTHUNDER,
+	finisher: ACTIONS.VERFLARE,
+}
+const DELAY_ACCELERATION_AVIALABLE_THRESHOLD = 4 // 4 seconds for 2 GCDs minus a 1 second window to activate before finisher
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: Resolved.

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+
+export default class MeleeCombos extends Module {
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: I intentionally did not try to combine the Combos.js module with the MeleeCombos module in this revision, the scope had already gotten larger than I really intended when I started with it.  Would be willing to take on collapsing it myself after this is done if desired.

File: src/parser/jobs/rdm/modules/MeleeCombos.js (hunk 3 above)
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: I think it was mostly my sheet anchor to windward, especially before I had the melee combo logic test cases to compare against.  Removed.

File: src/parser/jobs/rdm/modules/MeleeCombos.js (hunk 4 above)
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: Resolved.

File: src/parser/jobs/rdm/modules/MeleeCombos.js (hunk 2 above)
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: The two alternatives I could come up with were to give special meaning to null (or a boolean value), or to return both finishers and give that special meaning... and since I wanted to display the sequence recommended for a delay (which meant an array anyway), returning both finishers seemed like the better alternative.

File: src/parser/jobs/rdm/modules/MeleeCombos.js [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+	static handle = 'meleecombos'
+	// Disable unused dependency detection because module requires gauge for the fabricated event but no other data that triggers the used dependency check
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: Removed.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 14]
```diff
@@ -44,30 +42,124 @@ export const SEVERITY_LOST_MANA = {
...
 const ENHANCED_SCATTER_44_GAIN = 10
-const MISSING_HARDCAST_MANA_VALUE = 11
-const MANAIFCATION_MULTIPLIER = 2
-const MANA_DONT_CAST_THRESHOLD = 96
-const FINISHER_GAIN = 21
-const MELEE_COMBO_COST = 80
+//const MISSING_HARDCAST_MANA_VALUE = 11  //removed with refactor to GaugeAction, change to damage type events prevents need for guessing at missing casts
+const MANAFICATION_MULTIPLIER = 2
+
+class GaugeAction {
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: I don't think it would improve readability or reusability to split it out, since the class was created mostly to avoid having an anonymous object for the mana object being added to the fabricated event and to group the mana calculation functions for readability.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 15]
```diff
@@ -359,7 +257,7 @@ export default class Gauge extends Module {
 			}))
 
 			this.suggestions.add(new TieredSuggestion({
-				icon: ACTIONS.VERFLARE.icon,
+				icon: ACTIONS.VERHOLY.icon,
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: Wanted to bring this consistent with other white mana recommendations, since it looked like a copy paste error in the original code to me.

File: src/parser/jobs/rdm/modules/MeleeCombos.js (hunk 1 above)
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/MeleeCombos.js]: Implemented.  I'm not sure I like the readability of this method versus the original, but it's not a huge difference either way.

File: src/parser/jobs/rdm/modules/Gauge.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/jobs/rdm/modules/Gauge.js]: Ah, I meant the properties. Classes should be `PascalCase`.

File: src/parser/jobs/rdm/modules/Gauge.js [hunk 16]
```diff
@@ -1,24 +1,22 @@
...
-import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import {TieredSuggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+import {i18nMark, Trans} from '@lingui/react'
 //import {getCooldownRemaining} from 'parser/core/modules/Cooldowns'
 //import {ActionLink} from 'components/ui/DbLink'
 //TODO: Should possibly look into different Icons for things in Suggestions
 
 //Mana Gains and Expenditures
-const MANA_GAIN = {
+export const MANA_GAIN = {
```
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: To me, the MANA_GAIN (and CAP and DIFFERENCE_THRESHOLD) constants make sense as properties of the gauge, I think it's more intuitive to leave them in that module and exported so MeleeCombos can reference them.

File: src/parser/jobs/rdm/modules/Gauge.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (AzariahAeroborn) [reply src/parser/jobs/rdm/modules/Gauge.js]: Resolved.  Also found and fixed missed camelCase properties in the combo object in MeleeCombos.
